### ISSUE TITLE
RDCatch: Segfault on playback

### ISSUE THREAD

**Issue by alexolivan** [ElvishArtisan/rivendell#703]
This issue or another very similar has been already discussed: RDCatch segfaulting due to a database error
Here's the segfault in the qt5 branch in Bullseye

```
Aug 16 23:30:00 debian kernel: [35475.018435] rdcatchd[55058]: segfault at 5610772104a4 ip 0000560c75f2f2e5 sp 00007ffe83a51090 error 4 in rdcatchd[560c75f1b000+1e000]
Aug 16 23:30:00 debian kernel: [35475.018449] Code: 78 01 00 00 5b 5d 41 5c 41 5d 41 5e 41 5f c3 66 0f 1f 44 00 00 e8 bb 41 ff ff 49 8b bc 24 50 07 00 00 2d 81 00 00 00 48 01 df <41> 83 bc 84 28 02 00 00 01 0f 84 6b ff ff ff 48 8b 05 dd 5c 01 00
Aug 16 23:30:00 debian mariadbd[665]: 2021-08-16 23:30:00 212 [Warning] Aborted connection 212 to db: 'Rivendell' user: 'rduser' host: 'localhost' (Got an error reading communication packets)
``` 

In 3.6 the trick was to disable MariaDb strict mode:

```
[mysqld]
sql_mode=IGNORE_SPACE,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION
```

This but seems to be not working in Bullseye/Qt5 anymore, although I bet I'm doing something wrong... I give up for today... tomorrow will try to further tweak MariaDB and guess what is missing.

**Comment by alexolivan** [comment]
I have compared settings from MariaDB 10.3.27 (last in Buster) with 10.5.11 (Bullseye) regarding sql_mode (to disable strict mode) and, apparently, in Bullseye the configuration is correctly loaded and, still, rdcatch segfaults, wich is something that worries me a lot.

I've enabled error log on MariaDB error but I still have not a clue on what's going on with the error shown:
```
2021-08-17 13:00:00 36 [Warning] Aborted connection 36 to db: 'Rivendell' user: 'rduser' host: 'localhost' (Got an error reading communication packets)
```

I think I should figure out how to emulate an actual SQL query as it is done by rdcatch, to troubleshoot MariaDB settings.

**Comment by ElvishArtisan** [comment]
> In 3.6 the trick was to disable MariaDb strict mode:

Yes, but in 4.x we *want* it to work in ```strict``` mode, so we need to understand (and then fix) whatever issues in Rivendell make it break in that mode.

**Comment by ElvishArtisan** [comment]
Segfaults strongly imply a problem in the Rivendell code. What exactly are you doing to trigger that fault?

**Comment by ElvishArtisan** [comment]
> root@debian:~# mysql Rivendell -e "select 'ID','START_TIME' from 'RECORDINGS' where ('IS_ACTIVE'='Y')AND(time_to_sec('START_TIME')>time_to_sec('13:29:59'))AND(TUE='Y')order by 'START_TIME';"
ERROR 1064 (42000) at line 1: You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near ''RECORDINGS' where ('IS_ACTIVE'='Y')AND(time_to_sec('START_TIME')>time_to_sec...' at line 1

Your identifier quoting is incorrect. Use backticks (`)(ASCII 0x2C), not apostrophes (')(ASCII 0x27) for quoting identifiers. Apostrophes are used to quote string literals.

Thus:
```select `ID`,`START_TIME` from `RECORDINGS` where (`IS_ACTIVE`='Y')AND(time_to_sec(`START_TIME`)>time_to_sec('13:29:59'))AND(`TUE`='Y')order by `START_TIME`;"```
works as expected.

Added some warning language to 'CODINGSTYLE' in b7e264e.

**Comment by alexolivan** [comment]
> Yes, but in 4.x we _want_ it to work in `strict` mode, so we need to understand (and then fix) whatever issues in Rivendell make it break in that mode.

mmmm... this is very interesting indeed. So I must undo my non-strict mode tweak on MariaDB and keep up the test... I should have to figure that earllier. So I'll do and let's see what happens.



> Segfaults strongly imply a problem in the Rivendell code. What exactly are you doing to trigger that fault?

Nothing fancy: schedule a playback. that's all. Same cart (GTS BBC pips) at o'clock time, every hour.
Since I'm in holidays these days I have not a physical machine for tests, I'm using a VM, so the sound card is just Jackd Interface.
Setting the pips was the very first I did (I love the 'radio feel' of those GTS pips) but I did'nt notice they weren't playing because I was messing wit other stuff and jackd not routed to actual sound output... but finally It became evident.

I feel this bug could be hard one to hunt.

**Comment by ElvishArtisan** [comment]
Are you still seeing this problem? I've not been able to reproduce it at all here.

**Comment by alexolivan** [comment]
Yes... definitely is... RDCatch is not working (at least for me) in Debian (mariaDB)... a very severe issue.
If it does for others, then its my MariaDB setup.

What scares me is that, in 3.6 it was solveable by switching to  non-strict mode, but with 4.X, it just does not work.
Will but perform further retries on console... and try to further tweak MariaDB settings.

**Comment by alexolivan** [comment]
Would it be possible to fool rivendell to use a manually gdb-ran rdcatchd instance? ... being a daemon, my guess is that, it may not simply be a running instance that automagically 'joins' the other rivendell daemon environment participants.
The end goal is to have  an rdcatchd backtrace as it segfaults when it has to play a cart on schedule.

Regards.

**Comment by ElvishArtisan** [comment]
That's good data! Thanks.

And, to answer your original question: you can run one of the services under a debugger by starting them each 'by hand' rather than using ```rdservice(8)```. The order in which they are started is important, that being:
- caed
- ripcd
- rdcatchd
- rdpadd
- rdpadengined
- rdvairplayd
- rdrepld (optional, needed only if a replicator has been configured for that host)

For most debugging tasks, all you need are the first three (caed, ripcd and rdcatchd). ```rdservice(8)``` also has a series of '--end-startup-after-&lt;service&gt;' switches that can be used to partially automate this process; see the ```rdservice(8)``` man page for details.

**Comment by ElvishArtisan** [comment]
> ```[New Thread 0x7ffff15c8700 (LWP 794332)]
QObject::connect: No such signal QTcpSocket::connectionClosed() in rdcatchd.cpp:465
[Thread 0x7ffff15c8700 (LWP 794332) exited]
QObject::connect: No such signal QTcpSocket::connectionClosed() in rdcatchd.cpp:465```

Fixed in 49d6b2a.

**Comment by ElvishArtisan** [comment]
> And, to my surprise... on the next scheduled playout, it segfaulted! so, I have a bt:

Unfortunately, it looks like you were running a binary build --i.e. from a package -- without the debuginfo packages loaded, so we don't have a full source trace, but from the line numbers recorded it looks like we may have a situation where we have two 'overlapping' playback events trying to happen at the same time. Does that make any sense in the context of your RDCatch events list?

**Comment by alexolivan** [comment]
Yep... I always run from deb packages (I should prepare a VM to test stuff from a 'make install' install)
I installed the rivendell-dbgsym package... gdb seemed to recognize it. Before that, gdb bt didn't show anything about rdcatchd.cpp 605
```
Reading symbols from rdcatchd...
Reading symbols from /usr/lib/debug/.build-id/68/991f462fb308050d842702a5c28a4f61321d0c.debug...
```
I have installed every rivendell-*-dbgsym package, just in case, but it seems it makes no difference.

I have also repeated the tests, ensuring no to simultaneous playbacks, no simultanous access to same cart, no simultanous output of same audio output... same result.
Maybe something is wrong at a more basic level, maybe at underlying Operating System level, at the end of the day, it is just a no such 'file or directory' error: ```at rdcatchd.cpp:605 605	rdcatchd.cpp: No such file or directory. ```

Today, while updating the packages with tonight nightly builded packages...I noticed that:
```
chown: invalid group: ‘root:syslog’
chown: invalid user: ‘syslog:adm’
chown: invalid user: ‘syslog:adm’
``` 

syslog and adm groups are well known groups for logging facilities stuff .... maybe something is broken there: Is rivendell trying to write to any log beyond syslog / messages? I start to feel there's some kind of Debian-Ubuntu OS-level difference causing some issues

**Comment by ElvishArtisan** [comment]
> Today, while updating the packages with tonight nightly builded packages...I noticed that:
>
> chown: invalid group: ‘root:syslog’
> chown: invalid user: ‘syslog:adm’
> chown: invalid user: ‘syslog:adm’
>
>syslog and adm groups are well known groups for logging facilities stuff .... maybe something is broken there: Is rivendell trying to write to any log beyond syslog / messages? I start to feel there's some kind of Debian-Ubuntu OS-level difference causing some issues

Interesting. Can you post the wider context in which that is taking place?

**Comment by alexolivan** [comment]
Hi Fred... sure... here's the full context.
It's some stage on package apt install / update. The problem is reported in console during the package setup stage, so I'm guessing some kind of post-install script:

```
root@debian:~# apt-get update
Get:1 https://repo.alexolivan.com/debian bullseye/rivendell/snapshots InRelease [8,072 B]
Hit:2 http://deb.debian.org/debian bullseye InRelease                                                                                                
Get:3 http://security.debian.org/debian-security bullseye-security InRelease [44.1 kB]                                  
Get:4 https://repo.alexolivan.com/debian bullseye/rivendell/snapshots/main amd64 Packages [30.0 kB]                     
Get:5 https://www.deb-multimedia.org bullseye InRelease [15.8 kB]
Get:6 https://www.deb-multimedia.org bullseye/main amd64 Packages [129 kB]
Fetched 227 kB in 1s (180 kB/s)   
Reading package lists... Done
root@lanserver0:~# apt-get upgrade
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Calculating upgrade... Done
The following packages will be upgraded:
  rivendell rivendell-dbgsym rivendell-importers rivendell-importers-dbgsym rivendell-pypad rivendell-select rivendell-select-dbgsym rivendell-webapi rivendell-webapi-dbgsym
  rivendell-webget rivendell-webget-dbgsym
11 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
Need to get 140 MB of archives.
After this operation, 40.1 MB of additional disk space will be used.
Do you want to continue? [Y/n] Y
Get:1 https://repo.alexolivan.com/debian bullseye/rivendell/snapshots/main amd64 rivendell-webget-dbgsym amd64 4.0.0beta2-1+git20210913-7654811 [624 kB]
Get:2 https://repo.alexolivan.com/debian bullseye/rivendell/snapshots/main amd64 rivendell-webget amd64 4.0.0beta2-1+git20210913-7654811 [259 kB]
Get:3 https://repo.alexolivan.com/debian bullseye/rivendell/snapshots/main amd64 rivendell-select-dbgsym amd64 4.0.0beta2-1+git20210913-7654811 [1,613 kB]
Get:4 https://repo.alexolivan.com/debian bullseye/rivendell/snapshots/main amd64 rivendell-select amd64 4.0.0beta2-1+git20210913-7654811 [285 kB]
Get:5 https://repo.alexolivan.com/debian bullseye/rivendell/snapshots/main amd64 rivendell-importers-dbgsym amd64 4.0.0beta2-1+git20210913-7654811 [1,203 kB]
Get:6 https://repo.alexolivan.com/debian bullseye/rivendell/snapshots/main amd64 rivendell-importers amd64 4.0.0beta2-1+git20210913-7654811 [294 kB]
Get:7 https://repo.alexolivan.com/debian bullseye/rivendell/snapshots/main amd64 rivendell-dbgsym amd64 4.0.0beta2-1+git20210913-7654811 [129 MB]
Get:8 https://repo.alexolivan.com/debian bullseye/rivendell/snapshots/main amd64 rivendell amd64 4.0.0beta2-1+git20210913-7654811 [6,186 kB]
Get:9 https://repo.alexolivan.com/debian bullseye/rivendell/snapshots/main amd64 rivendell-pypad all 4.0.0beta2-1+git20210913-7654811 [263 kB]
Get:10 https://repo.alexolivan.com/debian bullseye/rivendell/snapshots/main amd64 rivendell-webapi-dbgsym amd64 4.0.0beta2-1+git20210913-7654811 [101 kB]
Get:11 https://repo.alexolivan.com/debian bullseye/rivendell/snapshots/main amd64 rivendell-webapi amd64 4.0.0beta2-1+git20210913-7654811 [268 kB]
Fetched 140 MB in 6s (24.5 MB/s)            
Reading changelogs... Done
(Reading database ... 120761 files and directories currently installed.)
Preparing to unpack .../00-rivendell-webget-dbgsym_4.0.0beta2-1+git20210913-7654811_amd64.deb ...
Unpacking rivendell-webget-dbgsym (4.0.0beta2-1+git20210913-7654811) over (4.0.0beta2-1+git20210911-49d6b2a) ...
Preparing to unpack .../01-rivendell-webget_4.0.0beta2-1+git20210913-7654811_amd64.deb ...
Unpacking rivendell-webget (4.0.0beta2-1+git20210913-7654811) over (4.0.0beta2-1+git20210911-49d6b2a) ...
Preparing to unpack .../02-rivendell-select-dbgsym_4.0.0beta2-1+git20210913-7654811_amd64.deb ...
Unpacking rivendell-select-dbgsym (4.0.0beta2-1+git20210913-7654811) over (4.0.0beta2-1+git20210911-49d6b2a) ...
Preparing to unpack .../03-rivendell-select_4.0.0beta2-1+git20210913-7654811_amd64.deb ...
Unpacking rivendell-select (4.0.0beta2-1+git20210913-7654811) over (4.0.0beta2-1+git20210911-49d6b2a) ...
Preparing to unpack .../04-rivendell-importers-dbgsym_4.0.0beta2-1+git20210913-7654811_amd64.deb ...
Unpacking rivendell-importers-dbgsym (4.0.0beta2-1+git20210913-7654811) over (4.0.0beta2-1+git20210911-49d6b2a) ...
Preparing to unpack .../05-rivendell-importers_4.0.0beta2-1+git20210913-7654811_amd64.deb ...
Unpacking rivendell-importers (4.0.0beta2-1+git20210913-7654811) over (4.0.0beta2-1+git20210911-49d6b2a) ...
Preparing to unpack .../06-rivendell-dbgsym_4.0.0beta2-1+git20210913-7654811_amd64.deb ...
Unpacking rivendell-dbgsym (4.0.0beta2-1+git20210913-7654811) over (4.0.0beta2-1+git20210911-49d6b2a) ...
Preparing to unpack .../07-rivendell_4.0.0beta2-1+git20210913-7654811_amd64.deb ...
Removed /etc/systemd/system/multi-user.target.wants/rivendell.service.
Unpacking rivendell (4.0.0beta2-1+git20210913-7654811) over (4.0.0beta2-1+git20210911-49d6b2a) ...
Preparing to unpack .../08-rivendell-pypad_4.0.0beta2-1+git20210913-7654811_all.deb ...
Unpacking rivendell-pypad (4.0.0beta2-1+git20210913-7654811) over (4.0.0beta2-1+git20210911-49d6b2a) ...
Preparing to unpack .../09-rivendell-webapi-dbgsym_4.0.0beta2-1+git20210913-7654811_amd64.deb ...
Unpacking rivendell-webapi-dbgsym (4.0.0beta2-1+git20210913-7654811) over (4.0.0beta2-1+git20210911-49d6b2a) ...
Preparing to unpack .../10-rivendell-webapi_4.0.0beta2-1+git20210913-7654811_amd64.deb ...
Unpacking rivendell-webapi (4.0.0beta2-1+git20210913-7654811) over (4.0.0beta2-1+git20210911-49d6b2a) ...
Setting up rivendell-pypad (4.0.0beta2-1+git20210913-7654811) ...
Setting up rivendell (4.0.0beta2-1+git20210913-7654811) ...
groupadd: group 'rivendell' already exists
useradd: user 'rivendell' already exists
groupadd: group 'pypad' already exists
useradd: user 'pypad' already exists
Created symlink /etc/systemd/system/multi-user.target.wants/rivendell.service → /lib/systemd/system/rivendell.service.
Synchronizing state of apache2.service with SysV service script with /lib/systemd/systemd-sysv-install.
Executing: /lib/systemd/systemd-sysv-install enable apache2
chown: invalid group: ‘root:syslog’
chown: invalid user: ‘syslog:adm’
chown: invalid user: ‘syslog:adm’
Setting up rivendell-webapi (4.0.0beta2-1+git20210913-7654811) ...
Setting up rivendell-importers (4.0.0beta2-1+git20210913-7654811) ...
Setting up rivendell-dbgsym (4.0.0beta2-1+git20210913-7654811) ...
Setting up rivendell-webget (4.0.0beta2-1+git20210913-7654811) ...
Setting up rivendell-importers-dbgsym (4.0.0beta2-1+git20210913-7654811) ...
Setting up rivendell-webapi-dbgsym (4.0.0beta2-1+git20210913-7654811) ...
Setting up rivendell-select (4.0.0beta2-1+git20210913-7654811) ...
Setting up rivendell-webget-dbgsym (4.0.0beta2-1+git20210913-7654811) ...
Setting up rivendell-select-dbgsym (4.0.0beta2-1+git20210913-7654811) ...
Processing triggers for mailcap (3.69) ...
Processing triggers for desktop-file-utils (0.26-1) ...
Processing triggers for hicolor-icon-theme (0.17-2) ...
Processing triggers for libc-bin (2.31-13) ...
Processing triggers for man-db (2.9.4-2) ...
```

Apparently Ubuntu does feature a syslog user and group whereas Debian does not.
adm group does exist both in Debian and Ubuntu.

What gets my attention is that rdcatcd.cpp line 605, as reported in bt as the origin of the segfault, just seems to perform a Play Deck Availability check... the crash seems to happen at that point.
I'm trying to ensure rdcatch config has an exclusive sound card output, but the 'No such file or directory' error makes me think it is trying to write to filesystem either to anin-exitant path or somewhere where it lacks permissions

**Comment by alexolivan** [comment]
> from the line numbers recorded it looks like we may have a situation where we have two 'overlapping' playback events trying to happen at the same time. Does that make any sense in the context of your RDCatch events list?

Definitely not... I have cleared all events, removed every deck and let just one playdeck. Then setup just a single, one shot, playback event... with same result.
Unfurtunatelly, there seems to be no libqt5core5a-dbgsym available and ?? dbg entries so correspond to Qt5 stuff... Would it help if I tried that to debug using a 'make install - like' installation?
moc_rdtimeengine.cpp being self generated turns my attention to rdtimeengine.cpp line 80

I'm going to try to schedule other types of events, etc... to see to what extent the thing is broken

**Comment by ElvishArtisan** [comment]
> I've also noticed that there's a dedicated /var/log/rivendell folder with a pair of log files 'debug' and 'operations' , along with a daily rotated set of files.
Every single log has 0 bytes... nothing is never, ever, landing there, which is somehow strange.
filesystem subtree /var/log/rivendell (and all its contents) is root:root owned with 640 perms.

Ha! That's _prima facie_ evidence that our ```rsyslogd(8)``` setup is borked on Ubuntu. Those files should contain the syslog messages from Rivendell (although the fact that you're seeing the 'debug' entries as well is strange; those should not have been enabled by default.

I'll have to spend some time with Ubuntu and track all this stuff down. (Probably not today, unfortunately).

**Comment by wpope** [comment]
I have been running rd 3.50 on ubuntu 18.04 for a year now.  I use rdcatchd extensively for downloads.  When downloads are intense , one every 2 or 3 minutes it seems that the system locks up.  ripcd, rdcatchd, mysql all seem to quit responding simultaneously .  Restarting mysql and rivendell daemons will get the system going again.  In my case it tends to be 2:30 in the morning.  Fortunately in am not running a airplay on this machine, and I am able to monitor heartbeat to detect and restart the process. I am not sure if it is related to this discussion.

**Comment by ElvishArtisan** [comment]
Possible of course, but there have been so many changes between v3.5.0 and the current qt5 branch (including a major rewrite for Qt5) that it'd be hard to trace the breadcrumbs.

**Comment by alexolivan** [comment]
> When downloads are intense , one every 2 or 3 minutes it seems that the system locks up
Mi first experience with rdcatch in 3.X lead to a rdcatchd crash, cause?: I mistakingly scheduled two simultaneous playouts in rdcatch... The impression is that, rdcatchd works processing events sequentially, and does nos support concurrency.
Maybe (just speculation) what you're experiencing could be due to some download/network bottleneck that lead to a situation where non-finished download events overlap with new scheduled ones, leading to a 'troublesome' concurrent event condition. 

Anyways, while I certainly prefer a coding style where the idea is to control every possible situation to avoid exceptions, rather that handling them, I've seen often, for those situations, a kind of 'God' process taking care of eventual respawn of critical/core processes

rdcatchd is, in my personal experince, the 'achiles heel' on Rivendell suite, since the user can (and I tend to do :-P), eventually, with its mere actions on a UI, crash a 'core' daemon.
It may be nice to have, for some selected user-interaction-exposed core daemons, such as rdcatchd or rdairplayd, to have some respawn/recover capability (maybe there already is! I'm not sure!)

**Comment by ElvishArtisan** [comment]
A little history here may be helpful. As originally conceived _ca._ 2002, RDCatch was a simple program recorder; with the original use case being recording feature-length programs off of analog satellite demods for later rebroadcast. The various other capabilites --switcher control, Internet operations, etc-- grew by accretion over the years. In the course of that, what was originally a rather straightforward codebase ramified into the patch-on-patch mess that you now see. And now the Jenga blocks are starting to topple.

So, I'm wondering if it isn't time to call in the B-52s and then start over. There'd be little or no change to the UI, but a completely new implementation of the service daemon. I'm thinking of something along the lines of a very simple main process that would be built around ```RDTimeEngine``` that would fork a separate process for each task. This would give some isolation between the core engine and the actual tasks. The other big change would be IPC; instead of the TCP-based control channel, we'd instead leverage the multicast notification infrastructure that was added back in v3.x.

**Comment by alexolivan** [comment]
> what was originally a rather straightforward codebase ramified into the patch-on-patch mess that you now see. And now the Jenga blocks are starting to topple.

That really explains alot.... Still, rdcatch/rdcatchd are incredible jobs.
In the automation systems I know, the wallclock-driven event management part is a key/core part of the software, and rdcatchd/rdcatch has nothing to envy to any of those, neither in functionality nor in stability (when properly used) if it wasnt designed as it is, it is a major achievement.

> So, I'm wondering if it isn't time to call in the B-52s and then start over.

From a non-experienced programmer perspective, this sounds as a major challenge... a ton of work and a lot of potential challenges to go through, really, really scaring!!!
From a more cold/theoretical perspective, I think such a decision could be the right one if that part of the code risks to keep lagging/drifting behind the overall project/code quality/design standards, because it reached where it was never engineered to be during its design/study phase. In that sense, the fact that those added use-cases turn to be of key/core relevance in the radio automation bussines-logic add some more weight. 

> I'm thinking of something along the lines of a very simple main process that would be built around `RDTimeEngine` that would fork a separate process for each task. This would give some isolation between the core engine and the actual tasks. The other big change would be IPC; instead of the TCP-based control channel

IMHO, provide that the decission is made to go for a major refactor/or totally new code, the only recomendation I can give to you, the ones who know the code, is to take your time in the plan/design phase... I think that the very early decisions are the momentous ones.

**Comment by ElvishArtisan** [comment]
> From a non-experienced programmer perspective, this sounds as a major challenge... a ton of work and a lot of potential challenges to go through, really, really scaring!!!
From a more cold/theoretical perspective, I think such a decision could be the right one if that part of the code risks to keep lagging/drifting behind the overall project/code quality/design standards, because it reached where it was never engineered to be during its design/study phase. In that sense, the fact that those added use-cases turn to be of key/core relevance in the radio automation bussines-logic add some more weight.

The UNIX programming culture includes some very strong DNA about writing simple programs that do one thing really well and then enabling them to communicate (sometimes referred to as the [Rule of Modularity](http://www.catb.org/~esr/writings/taoup/html/ch01s06.html#id2877537) and the [Rule of Composition](http://www.catb.org/~esr/writings/taoup/html/ch01s06.html#id2877684)). One of the many Good Things that tend to come from this is a greater willingness to abandon failed approaches in favor of re-engineered solutions based upon an improved understanding of the problem domain. The "scrap and rebuild" paradigm is actually fairly common in that world.

**Comment by ElvishArtisan** [comment]
Fixed in 2c74a1b. Please test!

**Comment by alexolivan** [comment]
Its fixed!!!! :-O

I didn't expect this one to get fixed in a very long time, being that the bug only afected Debian, not Ubuntu!
Also I recall your idea of re-building rdcatch from the ground-up .... so I guess you found something in the code doing something that could potentially spread to Ubuntu in the future.

Anyways... this is a milestone for Rivendell in Debian :-D

Clossing here... Thank you so much!
Cheers
